packages/core/src/browser/preferences/preference-provider.ts

```diff
@@ -41,20 +43,35 @@ export class PreferenceProvider implements Disposable {
         this.toDispose.dispose();
     }
 
-    protected fireOnDidPreferencesChanged(): void {
-        this.onDidPreferencesChangedEmitter.fire(undefined);
+    get<T>(preferenceName: string, resourceUri?: string): T | undefined {
+        const value = this.getPreferences(resourceUri)[preferenceName];
+        if (value !== undefined && value !== null) {
+            return value;
+        }
     }
 
-    getPreferences(): { [p: string]: any } {
-        return [];
+    getPreferences(resourceUri?: string): { [p: string]: any } {
+        return {};
     }
 
-    setPreference(key: string, value: any): Promise<void> {
+    setPreference(key: string, value: any, resourceUri?: string): Promise<void> {
         return Promise.resolve();
     }
 
     /** See `_ready`.  */
     get ready() {
         return this._ready.promise;
     }
+
+    canProvide(preferenceName: string, resourceUri?: string): { priority: number, provider: PreferenceProvider } {
+        return { priority: -1, provider: this };
+    }
+
+    getDomain(): string[] {
```

Comment:
What is domain?

Reply:
`domain` is where the preference changes could take effect. for example, the domain of the workspace-preferences-provider is the folders in that workspace. to be more specific, it is `["file:///A", "file:///B"]` in a workspace that has folder "A" and folder "B".